(docs): add quarkus java example

kameshsampath opened the pull request:
Add [QuarkusIO](https://quarkus.io) java example 

Fixes #1033

## Proposed Changes

- Add  helloworld java using  [QuarkusIO](https://quarkus.io) to serving samples
- The demo also has knative build based approach as well traditional java based approach

kameshsampath commented:
@samodell  - please review this and let me know if you require further changes.

kameshsampath commented:
I signed it!

samodell commented:
Hi @kameshsampath  -- can you update this PR to contribute your Quarkus sample to the community samples folder? https://github.com/knative/docs/tree/master/community/samples

kameshsampath commented:
Sure @samodell, will update and let you know. Thanks.

kameshsampath commented:
@samodell  - refactored to new folder. Please check and let me know if further changes needed.

@saturnism  - mind running the sample locally to test if all works for you ?

samodell commented:
/ok-to-test

grayside commented on community/samples/serving/helloworld-java-quarkus/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+RUN  /usr/local/bin/entrypoint-run.sh mvn -DskipTests clean package -Pnative 
+
+FROM registry.fedoraproject.org/fedora-minimal
```
Nice to see some more container registries. For minimal points of failure, is fedora-minimal available on quay.io? Or perhaps this could be moved to use Debian or Alpine from Docker Hub?

saturnism replied:
I don't think that's in our control, since this is a community contribution, and rh have already use quay.io to host their image.  I don't think we'd move this else where.

grayside commented on community/samples/serving/helloworld-java-quarkus/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+FROM registry.fedoraproject.org/fedora-minimal
+COPY --from=nativebuilder /project/target/helloworld-java-quarkus-runner /application
+EXPOSE 8080
```
No need to expose 8080. The port bound by knative applications is determined at runtime by the $PORT environment variable.

kameshsampath replied:
done

grayside commented on community/samples/serving/helloworld-java-quarkus/scripts/buildah.sh:
```diff
@@ -0,0 +1,55 @@
+buildah config --entrypoint '["/deployment/application"]'  $containerID
+buildah config --cmd "\-Dquarkus.http.host=0.0.0.0" $containerID
+
+buildah config --author "devx@redhat.com" --created-by "devx@redhat.com" --label Built-By=buildah $containerID
```
Should these @redhat.com email addresses be kept?

kameshsampath replied:
this script is no longer needed, removed it

kameshsampath commented:
/assign @rgregg

saturnism commented:
> So there's good news and bad news.
> 
> 👍 The good news is that everyone that needs to sign a CLA (the pull request submitter and all commit authors) have done so. Everything is all good there.
> 
> 😕 The bad news is that it appears that one or more commits were authored or co-authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.
> 
> _Note to project maintainer: This is a terminal state, meaning the `cla/google` commit status will not change from this state. It's up to you to confirm consent of all the commit author(s), set the `cla` label to `yes` (if enabled on your project), and then merge this pull request when appropriate._
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fknative%2Fdocs%2Fpull%2F1047) for more info**.

it's okay for my commit to be contributed to this project

grayside commented:
/lgtm

Follow-ups from @saturnism in follow-up PR.

RichieEscarez commented:
Are the followup PRs going to include some step-by-step instructions?

RichieEscarez commented:
Backtracking the linked items, I found that this PR is based on porting over something from https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/0.4.0/05-build/build.html.

Is the intention that users follow those instructions and use this sample code? That document links to other code samples so using it with this code sample will be confusing for users if/when they follow step-by-step. 

Can we provide some clear instructions for those who use this code sample (either there or along with these files)?

kameshsampath commented:
> Backtracking the linked items, I found that this PR is based on porting over something from https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/0.4.0/05-build/build.html.
> 
@rgregg - yes few items I ported from my tutorial https://redhat-developer-demos.github.io/knative-tutorial, but this is just to run a helloworld Quarkus alone and not all other build  related instruction. 

> Is the intention that users follow those instructions and use this sample code? That document links to other code samples so using it with this code sample will be confusing for users if/when they follow step-by-step.
> 
> Can we provide some clear instructions for those who use this code sample (either there or along with these files)?

Sorry I did not follow what you mean here  ? Can you please let me know ?

RichieEscarez commented:
> > Backtracking the linked items, I found that this PR is based on porting over something from https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/0.4.0/05-build/build.html.
> 
> @rgregg - yes few items I ported from my tutorial https://redhat-developer-demos.github.io/knative-tutorial, but this is just to run a helloworld Quarkus alone and not all other build related instruction. But it copy paste and modified for this sample.
> 
> > Is the intention that users follow those instructions and use this sample code? That document links to other code samples so using it with this code sample will be confusing for users if/when they follow step-by-step.
> > Can we provide some clear instructions for those who use this code sample (either there or along with these files)?
> 
> I don't think I added any references here other than the builder image repo.

Sorry for the confusion @kameshsampath, some how I managed to scroll past the README and thought it was missing. Ill start my review now.

kameshsampath commented:
> > > Backtracking the linked items, I found that this PR is based on porting over something from https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/0.4.0/05-build/build.html.
> > 
> > 
> > @rgregg - yes few items I ported from my tutorial https://redhat-developer-demos.github.io/knative-tutorial, but this is just to run a helloworld Quarkus alone and not all other build related instruction. But it copy paste and modified for this sample.
> > > Is the intention that users follow those instructions and use this sample code? That document links to other code samples so using it with this code sample will be confusing for users if/when they follow step-by-step.
> > > Can we provide some clear instructions for those who use this code sample (either there or along with these files)?
> > 
> > 
> > I don't think I added any references here other than the builder image repo.
> 
> Sorry for the confusion @kameshsampath, some how I managed to scroll past the README and thought it was missing. Ill start my review now.

No worries @rgregg - thanks taking it up.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+
+### Using Knative Build
+
+The sources also has a knative build file called `quarkus-build.yaml`, that will perform docker less container image build using [buildah](https://buildah.io)
```
```suggestion
To use Knative Build, you add your Docker Hub information to the `quarkus-build.yaml` file.
```

Moved optional info about build to the Before you begin section.

RichieEscarez replied:
There is no mention of installing Knative Build until now, which is late in the flow and forces the user to go back and install the Build component if in fact thats what they prefer.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+
+### Using Docker
+
+Use Docker to build the sample code into a container. To build and push with
```
```suggestion
To use Docker to locally build your container image and and push to Docker Hub, run the following commands with your Docker Hub username: 
```

RichieEscarez replied:
Add this to clarify that this is optional.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+### Using Docker
+
+Use Docker to build the sample code into a container. To build and push with
+Docker Hub, run these commands replacing `{username}` with your Docker Hub
```
```suggestion
```

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+
+Use Docker to build the sample code into a container. To build and push with
+Docker Hub, run these commands replacing `{username}` with your Docker Hub
+username:
```
```suggestion
```

kameshsampath commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+It uses an ENV variable `MESSAGE_PREFIX`(defaults to value `Hello`) and prints a message of format
+`${MESSAGE_PREFIX} Knative World!`.
+
+## Prerequisites
```
I have added the needed deps inside the builder image.

kameshsampath commented:
@rgregg  - thanks for your comments. Fixed them in last commit. Let me know if you require any further updates

saturnism commented:
/ping @RichieEscarez for review

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/.dockerignore:
```diff
@@ -0,0 +1,10 @@
+README.md
```
Should the _index.md also be listed here?

saturnism replied:
not critical, but added.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,452 @@
+
+## Before you begin
+
+You must meet the following requirements to run this sample:
```
Please see open comments/suggestions from first review. 
- Is both Knative Build and Docker required?
- The suggested edits make the list parallel and highlight options between docker or Knative Build

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,452 @@
+---
```
Delete lines 1-9 -> the _index.md file "includes" this readme during the build and the frontmatter nor the readfile statement should be here

saturnism replied:
removed

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+# Hello World - Quarkus Java sample
+
+A simple JAX-RS REST API application written in Java using [QuarkusIO](https://quarkus.io/).
```
Consider applying these suggestions to add clarity and target the subject (as "It" in the second sentence might refer to QuarkusIO).

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+folder) you're ready to build and deploy the sample app.
+
+We will be using hub.docker.com as the container registry, if you dont have an account yet have one created before proceeding further.
+
```
Are these both required?

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,452 @@
+generally more useful if you build them step-by-step. The following instructions
+recreate the source files from this folder.
+
+1. From the console, create a new empty web project using the maven archetype
```
For each step, we should be clear about where you want a file created. Is this at the root? Otherwise users are forced to look at and copy the repo.

saturnism replied:
what does it mean "is this at the root"? current verbiage seems to be inline w/ other examples, e.g., https://github.com/knative/docs/tree/master/docs/serving/samples/hello-world/helloworld-java-spring#recreating-the-sample-code

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
````diff
@@ -0,0 +1,452 @@
+    -Dpath="/"
+   ```
+
+1. Update the `GreetingResource` class in
````
I think we want the user to create this file too? We should first say create that file, state the location, and filename. Then, we have set the context and its clear what class is referred to.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
````diff
@@ -0,0 +1,452 @@
+
+   ```
+
+1. Add a new class the `GreetingService` class in
````
Same comment as above.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
````diff
@@ -0,0 +1,452 @@
+   ENTRYPOINT [ "/deployments/run-java.sh" ]
+   ```
+
+   If you want to build Quarkus native image, then copy the following code block in to file called `Dockerfile.native`
````
This is not a standalone element in the list and starts with "If", does that mean its optional?

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,452 @@
+
+   Go to `http://localhost:8080/` to see your `Hello World!` message.
+
+1. In your project directory, create a file named `Dockerfile` and copy the code
```
Where in the project directory? At root?

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
````diff
@@ -0,0 +1,452 @@
+   </project>
+   ```
+
+1. Run the application locally:
````
This should be a new section because its no longer about "recreating the code". Maybe "Locally testing your sample"

saturnism replied:
this is also inline w/ other samples w/ precedence. https://github.com/knative/docs/tree/master/docs/serving/samples/hello-world/helloworld-java-spring#recreating-the-sample-code

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,452 @@
+
+   Go to `http://localhost:8080/` to see your `Hello World!` message.
+
+1. In your project directory, create a file named `Dockerfile` and copy the code
```
Should all of these steps be moved before "locally testing your sample"? We should organize the content into distinct steps/sections so that its clear what the goal/purpose is at each stage in the guide.

RichieEscarez replied:
If the user chooses Knative Build, what would those steps be?

saturnism replied:
removed knative build section. that should prob go into knative build-template separately.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+
+## Prerequisites
+
+- A Kubernetes cluster with Knative installed. Follow the
```
Consider applying this set of edits to make the list parallel and highlight that Knative Build is an optional requirement.

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,419 @@
+- A Kubernetes cluster with Knative installed. Follow the
+  [installation instructions](https://github.com/knative/docs/blob/master/docs/install/README.md)
+  if you need to create one.
+- [Docker](https://www.docker.com) installed and running on your local machine,
```
revised this to clarify that docker is needed regardless, since you must push your container image to Docker Hub

RichieEscarez commented on community/samples/serving/helloworld-java-quarkus/README.md:
```diff
@@ -0,0 +1,452 @@
+- A Kubernetes cluster with Knative installed. Follow the
+  [installation instructions](https://github.com/knative/docs/blob/master/docs/install/README.md)
+  if you need to create one.
+- [Docker](https://www.docker.com) installed and running on your local machine,
```
Reordering steps to highlight Knative Build as optional (and that Docker is required to push to Docker Hub):
```suggestion
- You must install [Docker](https://www.docker.com) and have a Docker Hub account to which 
   you are able to upload your sample's container image.
- You can choose to use Knative Build or use Docker to build your sample's container image.
   - If you choose to use the [Knative Build component](../../../../docs/build/installing-build-component.md), 
     you must ensure that the Knative Build component is also installed in the cluster where you installed 
     Knative Serving. This method uses the [buildah](https://buildah.io) tool.
```

saturnism replied:
removed knative build section.

saturnism commented:
/ping @RichieEscarez updated w/ your comments. ptal.

grayside commented on community/samples/serving/helloworld-java-quarkus/Dockerfile.native:
```diff
@@ -0,0 +1,16 @@
+COPY --from=builder /project/target/helloworld-java-quarkus-runner /app
+
+ENTRYPOINT [ "/app" ]
+CMD [ "-Dquarkus.http.host=0.0.0.0","-Dquarkus.http.port=8080" ]
```
Knative apps should listen to the runtime environment variable $PORT for the port directive.

saturnism replied:
fixed. added instruction to configure PORT via env var.

grayside commented on community/samples/serving/helloworld-java-quarkus/Dockerfile.native:
```diff
@@ -0,0 +1,16 @@
+FROM quay.io/rhdevelopers/quarkus-java-builder:graal-1.0.0-rc15 as builder
```
Non-blocking nit: Some IDEs recognize *.Dockerfile as a Dockerfile but not Dockerfile.*.

grayside commented on community/samples/serving/helloworld-java-quarkus/src/main/docker/Dockerfile.native:
```diff
@@ -0,0 +1,22 @@
+WORKDIR /work/
+COPY target/*-runner /work/application
+RUN chmod 775 /work
+EXPOSE 8080
```
$PORT env var here too?

saturnism replied:
these files are not needed, removed.

grayside commented on community/samples/serving/helloworld-java-quarkus/Dockerfile.native:
```diff
@@ -0,0 +1,17 @@
+
+RUN /usr/local/bin/entrypoint-run.sh mvn -DskipTests clean package -Pnative
+
+RUN ls -lh /project/target
```
Stray debug code?

saturnism replied:
my mistake - good catch!

RichieEscarez commented:
Edited the first part to ensure the user can get up and running, otherwise left the content as is.

Going to approve this to get back to 0.6 release items. @grayside or @saturnism this needs an LGTM to merge (please)

saturnism commented:
Thank you!
